// @iomechs/angular-tags-input/esm2015/lib/tag/tag.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
let TagComponent = class TagComponent {
    constructor() {
        this.tagClass = 'default-tag';
        this.closeClicked = new EventEmitter();
    }
    ngOnInit() {
        if (!this.tagItemTemplate) { // if there's no template provided, assign the default one
            this.tagItemTemplate = this.defaultTagItemTemplate;
        }
        this.context = {
            item: this.tagItem,
            config: this.config,
            tagClass: this.tagClass,
            closeClicked: this.closeClicked
        };
    }
};
tslib_1.__decorate([
    Input()
], TagComponent.prototype, "config", void 0);
tslib_1.__decorate([
    Input()
], TagComponent.prototype, "tagItem", void 0);
tslib_1.__decorate([
    Input()
], TagComponent.prototype, "tagClass", void 0);
tslib_1.__decorate([
    Input()
], TagComponent.prototype, "tagItemTemplate", void 0);
tslib_1.__decorate([
    Output()
], TagComponent.prototype, "closeClicked", void 0);
tslib_1.__decorate([
    ViewChild('defaultTagItemTemplate', { static: true })
], TagComponent.prototype, "defaultTagItemTemplate", void 0);
TagComponent = tslib_1.__decorate([
    Component({
        selector: 'ti-tag',
        template: "\n<ng-container\n *ngTemplateOutlet=\"tagItemTemplate; context: context;\">\n</ng-container>\n\n<!-- Default Tag Options Template -->\n<ng-template #defaultTagItemTemplate let-item=\"item\" let-tagClass=\"tagClass\" let-config=\"config\" let-closeClicked=\"closeClicked\">\n  <div class=\"angular-tag\" [ngClass]=\"tagClass + ''\">\n    <div class=\"angular-tag__text\" *ngIf=\"config\">\n      {{item[config.displayProperty]}}\n    </div>\n    <button type=\"button\" class=\"angular-tag__close-btn\" (click)=\"closeClicked.emit(item)\">\n      x\n    </button>\n  </div>\n</ng-template>\n\n\n",
        styles: [".angular-tag{display:flex;align-items:center}.angular-tag__text{margin-right:10px;flex:1}.default-tag .angular-tag__text{color:#fff}.angular-tag__close-btn{cursor:pointer}.default-tag .angular-tag__close-btn{padding-left:0;padding-right:0;background:0 0;border:none;color:#fff}.default-tag .angular-tag__close-btn:hover{color:#dcdcdc}.angular-tag.default-tag{background:#344860;padding:6px 10px;transition:.5s}.angular-tag.default-tag:hover{background:#333}"]
    })
], TagComponent);
export { TagComponent };
