// @iomechs/angular-tags-input/esm2015/lib/tag-input/tag-input.component.js
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { FormGroup, FormControl } from '@angular/forms';
import { debounceTime, distinctUntilChanged } from 'rxjs/operators';
let TagInputComponent = class TagInputComponent {
    constructor() {
        this.inputClass = 'default';
        this.disabled = false;
        this.valueChanged = new EventEmitter();
        this.inputFocused = new EventEmitter();
        this.inputBlurred = new EventEmitter();
        this.tagEntered = new EventEmitter();
        this.tagInputForm = new FormGroup({
            tagInputVal: new FormControl('')
        });
    }
    ngOnInit() {
        this.tagInputForm.get('tagInputVal')
            .valueChanges
            .pipe(debounceTime(400), distinctUntilChanged())
            .subscribe((value) => {
            this.lastValueEmitted = value;
            this.valueChanged.emit(value);
        });
    }
    ngOnChanges(changes) {
        if (changes.disabled !== undefined) {
            if (changes.disabled.currentValue === true) {
                this.tagInputForm.get('tagInputVal').disable();
            }
            else {
                this.tagInputForm.get('tagInputVal').enable();
            }
        }
    }
    /**
     * @author Ahsan Ayaz
     * @desc Emits the input focused event with the current search term value
     */
    emitInputFocused() {
        if (this.disabled) {
            return;
        }
        const searchTerm = this.tagInputForm.get('tagInputVal').value;
        this.inputFocused.emit(searchTerm);
        if (this.lastValueEmitted !== searchTerm) {
            this.valueChanged.emit(searchTerm);
            this.lastValueEmitted = searchTerm;
        }
    }
    /**
     * @author Ahsan Ayaz
     * @desc When the user presses enter key after entering the tag name
     * @param inputEl - Reference to the input element so we can fetch the value
     */
    onEnterKeyPress() {
        if (this.config.onlyFromAutoComplete) {
            return;
        }
        const tagName = this.inputEl.nativeElement.value;
        this.tagEntered.emit({
            [this.config.displayProperty]: tagName,
            [this.config.identifier]: tagName.toLowerCase(),
            newTag: true
        });
        this.resetInput();
    }
    /**
     * @author Ahsan Ayaz
     * @desc Resets the input value
     */
    resetInput() {
        this.inputEl.nativeElement.value = '';
    }
    emitInputBlurred($event) {
        this.inputBlurred.emit($event);
    }
};
tslib_1.__decorate([
    Input()
], TagInputComponent.prototype, "inputClass", void 0);
tslib_1.__decorate([
    Input()
], TagInputComponent.prototype, "config", void 0);
tslib_1.__decorate([
    Input()
], TagInputComponent.prototype, "disabled", void 0);
tslib_1.__decorate([
    Output()
], TagInputComponent.prototype, "valueChanged", void 0);
tslib_1.__decorate([
    Output()
], TagInputComponent.prototype, "inputFocused", void 0);
tslib_1.__decorate([
    Output()
], TagInputComponent.prototype, "inputBlurred", void 0);
tslib_1.__decorate([
    Output()
], TagInputComponent.prototype, "tagEntered", void 0);
tslib_1.__decorate([
    ViewChild('inputEl', { static: true })
], TagInputComponent.prototype, "inputEl", void 0);
TagInputComponent = tslib_1.__decorate([
    Component({
        selector: 'ti-tag-input',
        template: "<div class=\"angular-tag-input\" [ngClass]=\"inputClass + ''\">\n  <form no-validate [formGroup]=\"tagInputForm\">\n    <input\n      #inputEl\n      [disabled]=\"disabled\"\n      (keyup.enter)=\"onEnterKeyPress()\"\n      [placeholder]=\"config?.placeholder\"\n      formControlName=\"tagInputVal\"\n      (click)=\"emitInputFocused()\"\n      (focus)=\"emitInputFocused()\"\n      (blur)=\"emitInputBlurred($event)\"\n      class=\"angular-tag-input__input\"\n      autocomplete=\"off\">\n  </form>\n</div>\n",
        styles: [".default .angular-tag-input__input{padding:6px 10px;border:none;width:98%;min-width:100px}.angular-tag-input input:focus{outline-width:0}"]
    })
], TagInputComponent);
export { TagInputComponent };
