// @iomechs/angular-tags-input/esm5/lib/keyboard-active-class/keyboard-active-class.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input, OnChanges, SimpleChanges, ElementRef, ChangeDetectorRef, OnInit } from '@angular/core';
var KeyboardActiveClassDirective = /** @class */ (function () {
    function KeyboardActiveClassDirective(el, cdRef) {
        this.el = el;
        this.cdRef = cdRef;
        this.isKeyboardActiveItem = false;
    }
    KeyboardActiveClassDirective.prototype.ngOnInit = function () {
        if (!this.tiKeyboardActiveClass) {
            console.warn('tiKeyboardActiveClass needs a class to apply. But no class was passed.');
        }
    };
    KeyboardActiveClassDirective.prototype.ngOnChanges = function (simpleChanges) {
        if (!simpleChanges.isKeyboardActiveItem) {
            return;
        }
        if (simpleChanges.isKeyboardActiveItem.currentValue !== simpleChanges.isKeyboardActiveItem.previousValue) {
            this.evaluateClassApplication();
        }
    };
    KeyboardActiveClassDirective.prototype.evaluateClassApplication = function () {
        if (this.isKeyboardActiveItem) {
            this.el.nativeElement.classList.add(this.tiKeyboardActiveClass);
            this.el.nativeElement.scrollIntoView(false);
        }
        else {
            this.el.nativeElement.classList.remove(this.tiKeyboardActiveClass);
        }
        this.cdRef.markForCheck();
    };
    KeyboardActiveClassDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    tslib_1.__decorate([
        Input()
    ], KeyboardActiveClassDirective.prototype, "isKeyboardActiveItem", void 0);
    tslib_1.__decorate([
        Input()
    ], KeyboardActiveClassDirective.prototype, "tiKeyboardActiveClass", void 0);
    KeyboardActiveClassDirective = tslib_1.__decorate([
        Directive({
            selector: '[tiKeyboardActiveClass]'
        })
    ], KeyboardActiveClassDirective);
    return KeyboardActiveClassDirective;
}());
export { KeyboardActiveClassDirective };
