// @iomechs/angular-tags-input/esm5/lib/dropdown/dropdown.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { ListKeyManager } from '@angular/cdk/a11y';
import { UP_ARROW, DOWN_ARROW, ENTER } from '@angular/cdk/keycodes';
import { KEY_CODES } from '../constants';
import { AngularTagsInputService } from '../angular-tags-input.service';
import { DropdownItemsFilterPipe } from '../dropdown-items-filter.pipe';
var DropdownComponent = /** @class */ (function () {
    function DropdownComponent(tagsInputService) {
        this.tagsInputService = tagsInputService;
        this.listItems = [];
        this.inputVal = '';
        this.itemAdded = new EventEmitter();
        this.itemClicked = new EventEmitter();
        this.dropdownItemsFilter = new DropdownItemsFilterPipe();
        this.identifierSeparator = '__';
        this.itemsMap = new Map();
    }
    DropdownComponent.prototype.ngOnInit = function () {
        if (!this.dropDownTemplate) {
            // if there's no template provided, assign the default one
            this.dropDownTemplate = this.defaultTagOptionTemplate;
        }
        this.context = {
            items: tslib_1.__spread(this.listItems),
            config: this.config,
            tagsLoading: this.tagsLoading,
            fns: {
                onItemClicked: this.onItemClicked.bind(this)
            }
        };
        this.keyboardEventsManager = new ListKeyManager(tslib_1.__spread(this.listItems));
        this.populateItemsMap(this.listItems);
        this.ddIdPrefix = this.getRandomString();
        this.tagsInputService.log(this.itemsMap, 'items populated initially');
    };
    DropdownComponent.prototype.ngOnChanges = function (changes) {
        if (changes.listItems && !changes.listItems.firstChange) {
            // if the list items change, update the context items (because they're not automatically updated)
            this.filterItems(this.inputVal, changes.listItems.currentValue);
            this.populateItemsMap(changes.listItems ? changes.listItems.currentValue : this.listItems);
        }
        if (changes.inputVal && !changes.inputVal.firstChange) {
            // if the list items change, update the context items (because they're not automatically updated)
            this.filterItems(changes.inputVal.currentValue);
            this.populateItemsMap(changes.listItems ? changes.listItems.currentValue : this.listItems);
        }
    };
    /**
     * @author Ahsan Ayaz
     * @desc Updates the items property for the context provided to the dropdown template
     * @param items - the list of items to be assigned
     */
    DropdownComponent.prototype.filterItems = function (searchTerm, items) {
        if (searchTerm === void 0) { searchTerm = this.inputVal; }
        if (items === void 0) { items = this.listItems; }
        this.context.items = tslib_1.__spread(this.dropdownItemsFilter.transform(items, this.config, searchTerm));
        this.ddIdPrefix = this.getRandomString();
    };
    DropdownComponent.prototype.getRandomString = function () {
        return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
    };
    DropdownComponent.prototype.populateItemsMap = function (items, prefix) {
        var _this = this;
        if (prefix === void 0) { prefix = null; }
        if (prefix === null) {
            prefix = '';
        }
        else {
            prefix += this.identifierSeparator;
        }
        items.map(function (item, index) {
            var newPrefix = prefix + index;
            item.tiIdentifier = newPrefix;
            _this.itemsMap[newPrefix] = item;
            if (item[_this.config.nestedTagProperty] && item[_this.config.nestedTagProperty].length) {
                _this.populateItemsMap(item[_this.config.nestedTagProperty], newPrefix);
            }
        });
    };
    /**
     * @author Ahsan Ayaz
     * @desc When an option is clicked from the options dropdown
     * @param item - item clicked
     */
    DropdownComponent.prototype.onItemClicked = function (item, $event) {
        if ($event === void 0) { $event = null; }
        if ($event) {
            $event.stopImmediatePropagation(); // for nested items
        }
        this.itemClicked.emit(item);
    };
    DropdownComponent.prototype.handleKeyUp = function (event) {
        event.stopImmediatePropagation();
        if (!this.keyboardEventsManager) {
            return;
        }
        var isKeyDown = this.isKeyDown(event);
        var isKeyUp = this.isKeyUp(event);
        var isKeyEnter = this.isKeyEnter(event);
        if (isKeyDown || isKeyUp) {
            // passing the event to key manager so we get a change fired
            this.setActiveElement(event);
        }
        else if (isKeyEnter &&
            this.keyboardEventsManager.activeItem) {
            this.itemClicked.emit(this.keyboardEventsManager.activeItem);
        }
    };
    DropdownComponent.prototype.isKeyDown = function (event) {
        // tslint:disable-next-line: deprecation
        return event.key === KEY_CODES.ARROW_DOWN || event.keyCode === DOWN_ARROW;
    };
    DropdownComponent.prototype.isKeyUp = function (event) {
        // tslint:disable-next-line: deprecation
        return event.key === KEY_CODES.ARROW_UP || event.keyCode === UP_ARROW;
    };
    DropdownComponent.prototype.isKeyEnter = function (event) {
        // tslint:disable-next-line: deprecation
        return event.key === KEY_CODES.ENTER || event.keyCode === ENTER;
    };
    DropdownComponent.prototype.setActiveElement = function (event) {
        var isKeyDown = this.isKeyDown(event);
        var isKeyUp = this.isKeyUp(event);
        var previousActiveItem = tslib_1.__assign({}, this.keyboardEventsManager.activeItem);
        if (isKeyDown) {
            this.setNextActiveElement(previousActiveItem, this.listItems);
        }
        else if (isKeyUp) {
            this.setPreviousActiveElement(previousActiveItem, this.listItems);
        }
    };
    DropdownComponent.prototype.setNextActiveElement = function (currentActiveItem, items) {
        var _this = this;
        var identifier = currentActiveItem.tiIdentifier;
        if (!identifier) {
            items[0].tiKeyboardActive = true;
            this.keyboardEventsManager.setFirstItemActive();
            return;
        }
        var keyIdentifierArr = identifier.split(this.identifierSeparator);
        var keyIdentifier;
        var treeItems;
        var index;
        keyIdentifier = keyIdentifierArr.length > 1 ?
            tslib_1.__spread(keyIdentifierArr).splice(0, keyIdentifierArr.length - 1).join(this.identifierSeparator) :
            keyIdentifierArr[0];
        treeItems = Object.keys(this.itemsMap).filter(function (key) { return new RegExp("^" + keyIdentifier + _this.identifierSeparator).test(key); });
        if (!treeItems.length) {
            treeItems = Object.keys(this.itemsMap).filter(function (key) { return new RegExp("^" + keyIdentifier + "$").test(key); });
        }
        index = treeItems.findIndex(function (id) { return id === identifier; });
        if (index === treeItems.length - 1) {
            this.setActiveElementRecursively(this.findNextParent(keyIdentifierArr), this.listItems);
        }
        else {
            this.setActiveElementRecursively(treeItems[index + 1], this.listItems);
        }
    };
    DropdownComponent.prototype.findNextParent = function (keyIdentifierArr) {
        var keyIdentifier;
        var treeItems;
        if (keyIdentifierArr.length === 1) {
            var nextIdentifier = "" + (+keyIdentifierArr[0] + 1);
            if (this.itemsMap[nextIdentifier]) {
                return nextIdentifier;
            }
            return keyIdentifierArr[0];
        }
        var nextIdArr = tslib_1.__spread(keyIdentifierArr).splice(0, keyIdentifierArr.length - 1);
        nextIdArr[nextIdArr.length - 1] = "" + ((+nextIdArr[nextIdArr.length - 1]) + 1);
        keyIdentifier = nextIdArr.join(this.identifierSeparator);
        treeItems = Object.keys(this.itemsMap).filter(function (key) { return new RegExp("^" + keyIdentifier).test(key); });
        if (treeItems.length) {
            return treeItems[0];
        }
        else {
            return this.findNextParent(tslib_1.__spread(keyIdentifierArr.splice(0, keyIdentifierArr.length - 1)));
        }
    };
    DropdownComponent.prototype.setPreviousActiveElement = function (currentActiveItem, items) {
        var _this = this;
        var identifier = currentActiveItem.tiIdentifier;
        if (!identifier) {
            items[items.length - 1].tiKeyboardActive = true;
            this.keyboardEventsManager.setActiveItem(items[items.length - 1]);
            return;
        }
        var keyIdentifierArr = identifier.split(this.identifierSeparator);
        var keyIdentifier;
        var treeItems;
        var index;
        keyIdentifier = keyIdentifierArr.length > 1 ?
            tslib_1.__spread(keyIdentifierArr).splice(0, keyIdentifierArr.length - 1).join(this.identifierSeparator) :
            keyIdentifierArr[0];
        treeItems = Object.keys(this.itemsMap).filter(function (key) { return new RegExp("^" + keyIdentifier + "(" + _this.identifierSeparator + ")?").test(key); });
        if (!treeItems.length) {
            treeItems = Object.keys(this.itemsMap).filter(function (key) { return new RegExp("^" + keyIdentifier + "$").test(key); });
        }
        index = treeItems.findIndex(function (id) { return id === identifier; });
        if (treeItems.length === 1 || index === 0) {
            this.setActiveElementRecursively(this.findPrevousParentLastChild(keyIdentifierArr, identifier), this.listItems);
        }
        else {
            this.setActiveElementRecursively(treeItems[index - 1], this.listItems);
        }
    };
    DropdownComponent.prototype.findPrevousParentLastChild = function (keyIdentifierArr, prevItemIdentifier) {
        var _this = this;
        var keyIdentifier;
        var isOnlyItem = keyIdentifierArr.length === 1;
        var prevIdArr = !!isOnlyItem ? tslib_1.__spread(keyIdentifierArr) :
            tslib_1.__spread(keyIdentifierArr).splice(0, keyIdentifierArr.length - 1);
        prevIdArr[prevIdArr.length - 1] = !!isOnlyItem ? "" + ((+prevIdArr[prevIdArr.length - 1]) - 1) : prevIdArr[prevIdArr.length - 1];
        keyIdentifier = prevIdArr.join(this.identifierSeparator);
        var treeItems = Object.keys(this.itemsMap).filter(function (key) { return new RegExp("^" + keyIdentifier + _this.identifierSeparator).test(key); });
        if (!treeItems.length) {
            treeItems = Object.keys(this.itemsMap).filter(function (key) { return new RegExp("^" + keyIdentifier + "$").test(key); });
        }
        if (treeItems.length) {
            if (treeItems.indexOf(prevItemIdentifier) !== -1) {
                return keyIdentifier;
            }
            return treeItems[treeItems.length - 1];
        }
        else if (keyIdentifierArr.length > 1) {
            return this.findPrevousParentLastChild(tslib_1.__spread(keyIdentifierArr.splice(0, keyIdentifierArr.length - 1)), prevItemIdentifier);
        }
        else {
            keyIdentifier = keyIdentifierArr.join(this.identifierSeparator);
            return keyIdentifier;
        }
    };
    DropdownComponent.prototype.setActiveElementRecursively = function (identifier, items) {
        for (var i = 0, len = items.length; i < len; ++i) {
            items[i].tiKeyboardActive = false;
            if (items[i].tiIdentifier === identifier) {
                this.keyboardEventsManager = new ListKeyManager(tslib_1.__spread(items));
                items[i].tiKeyboardActive = true; // select next item
                this.keyboardEventsManager.setActiveItem(items[i]);
            }
            if (items[i][this.config.nestedTagProperty] && items[i][this.config.nestedTagProperty].length) {
                this.setActiveElementRecursively(identifier, items[i][this.config.nestedTagProperty]);
            }
        }
    };
    DropdownComponent.ctorParameters = function () { return [
        { type: AngularTagsInputService }
    ]; };
    tslib_1.__decorate([
        Input()
    ], DropdownComponent.prototype, "config", void 0);
    tslib_1.__decorate([
        Input()
    ], DropdownComponent.prototype, "listItems", void 0);
    tslib_1.__decorate([
        Input()
    ], DropdownComponent.prototype, "inputVal", void 0);
    tslib_1.__decorate([
        Input()
    ], DropdownComponent.prototype, "dropDownTemplate", void 0);
    tslib_1.__decorate([
        Input()
    ], DropdownComponent.prototype, "tagsLoading", void 0);
    tslib_1.__decorate([
        Input()
    ], DropdownComponent.prototype, "keyPress", void 0);
    tslib_1.__decorate([
        Output()
    ], DropdownComponent.prototype, "itemAdded", void 0);
    tslib_1.__decorate([
        Output()
    ], DropdownComponent.prototype, "itemClicked", void 0);
    tslib_1.__decorate([
        ViewChild('defaultTagOptionTemplate', { static: true })
    ], DropdownComponent.prototype, "defaultTagOptionTemplate", void 0);
    DropdownComponent = tslib_1.__decorate([
        Component({
            selector: 'ti-dropdown',
            template: "\n<ng-container\n *ngTemplateOutlet=\"dropDownTemplate; context: context\">\n</ng-container>\n\n<!-- Default Tag Options Template -->\n<ng-template #defaultTagOptionTemplate let-items=\"items\" let-config=\"config\">\n  <div class=\"angular-tags-dropdown\">\n    <div class=\"angular-tags-dropdown__list\">\n      <ng-container *ngFor=\"let item of items, let i = index\">\n        <div\n          class=\"angular-tags-dropdown__list__item\"\n          (click)=\"onItemClicked(item)\"\n          tiKeyboardActiveClass=\"angular-tags-dropdown__list__item--active\"\n          [isKeyboardActiveItem]=\"item.tiKeyboardActive\"\n          >\n          {{item[config.displayProperty]}}\n        </div>\n      </ng-container>\n    </div>\n  </div>\n</ng-template>\n",
            styles: [":host{display:block}:host .angular-tags-dropdown{padding:8px 10px}:host .angular-tags-dropdown__list__item{padding:8px 10px;cursor:pointer;transition:.5s}:host .angular-tags-dropdown__list__item:hover{background:#dcdcdc}:host .angular-tags-dropdown__list__item--active{background:#333;color:#fff}"]
        })
    ], DropdownComponent);
    return DropdownComponent;
}());
export { DropdownComponent };
